File: app/src/main/java/com/elementary/tasks/reminder/models/Reminder.java
package com.elementary.tasks.reminder.models;

import com.elementary.tasks.core.interfaces.RecyclerInterface;
import com.elementary.tasks.core.utils.SuperUtil;
import com.elementary.tasks.core.utils.TimeUtil;
import com.elementary.tasks.creators.fragments.ReminderInterface;
import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Random;
import java.util.UUID;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class Reminder implements RecyclerInterface {

    private static final String TAG = "Reminder";

    public static final int REMINDER = 0;
    public static final int SHOPPING = 1;

    public static final int BY_DATE = 10;
    public static final int BY_DATE_CALL = 11;
    public static final int BY_DATE_SMS = 12;
    public static final int BY_DATE_APP = 13;
    public static final int BY_DATE_LINK = 14;
    public static final int BY_DATE_SHOP = 15;
    public static final int BY_DATE_EMAIL = 16;
    public static final int BY_TIME = 20;
    public static final int BY_WEEK = 30;
    public static final int BY_WEEK_CALL = 31;
    public static final int BY_WEEK_SMS = 32;
    public static final int BY_LOCATION = 40;
    public static final int BY_LOCATION_CALL = 41;
    public static final int BY_LOCATION_SMS = 42;
    public static final int BY_SKYPE = 50;
    public static final int BY_SKYPE_CALL = 51;
    public static final int BY_SKYPE_VIDEO = 52;
    public static final int BY_MONTH = 60;
    public static final int BY_MONTH_CALL = 61;
    public static final int BY_MONTH_SMS = 62;
    public static final int BY_OUT = 70;
    public static final int BY_OUT_CALL = 71;
    public static final int BY_OUT_SMS = 72;
    public static final int BY_PLACES = 80;
    public static final int BY_PLACES_CALL = 81;
    public static final int BY_PLACES_SMS = 82;
    public static final int BY_DAY_OF_YEAR = 90;
    public static final int BY_DAY_OF_YEAR_CALL = 91;
    public static final int BY_DAY_OF_YEAR_SMS = 92;

    @SerializedName("summary")
    private String summary;
    @SerializedName("noteId")
    private String noteId;
    @SerializedName("reminderType")
    private int reminderType;
    @SerializedName("groupUuId")
    private String groupUuId;
    @SerializedName("uuId")
    private String uuId;
    @SerializedName("eventTime")
    private String eventTime;
    @SerializedName("startTime")
    private String startTime;
    @SerializedName("eventCount")
    private long eventCount;
    @SerializedName("color")
    private int color;
    @SerializedName("delay")
    private int delay;
    @SerializedName("vibrate")
    private boolean vibrate;
    @SerializedName("repeatNotification")
    private boolean repeatNotification;
    @SerializedName("notifyByVoice")
    private boolean notifyByVoice;
    @SerializedName("awake")
    private boolean awake;
    @SerializedName("unlock")
    private boolean unlock;
    @SerializedName("exportToTasks")
    private boolean exportToTasks;
    @SerializedName("exportToCalendar")
    private boolean exportToCalendar;
    @SerializedName("useGlobal")
    private boolean useGlobal;
    @SerializedName("from")
    private String from;
    @SerializedName("to")
    private String to;
    @SerializedName("hours")
    private List<Integer> hours = new ArrayList<>();
    @SerializedName("fileName")
    private String fileName;
    @SerializedName("melodyPath")
    private String melodyPath;
    @SerializedName("volume")
    private int volume;
    @SerializedName("dayOfMonth")
    private int dayOfMonth;
    @SerializedName("monthOfYear")
    private int monthOfYear;
    @SerializedName("repeatInterval")
    private long repeatInterval;
    @SerializedName("repeatLimit")
    private int repeatLimit;
    @SerializedName("after")
    private long after;
    @SerializedName("weekdays")
    private List<Integer> weekdays = new ArrayList<>();
    @SerializedName("type")
    private int type;
    @SerializedName("target")
    private String target;
    @SerializedName("subject")
    private String subject;
    @SerializedName("attachmentFile")
    private String attachmentFile;
    @SerializedName("attachmentFiles")
    private List<String> attachmentFiles = new ArrayList<>();
    @SerializedName("auto")
    private boolean auto;
    @SerializedName("places")
    private List<Place> places = new ArrayList<>();
    @SerializedName("shoppings")
    private List<ShopItem> shoppings = new ArrayList<>();
    @SerializedName("uniqueId")
    private int uniqueId;
    @SerializedName("isActive")
    private boolean isActive;
    @SerializedName("isRemoved")
    private boolean isRemoved;
    @SerializedName("isNotificationShown")
    private boolean isNotificationShown;
    @SerializedName("isLocked")
    private boolean isLocked;
    @SerializedName("duration")
    private long duration;
    @SerializedName("remindBefore")
    private long remindBefore;

    public static boolean isBase(int type, int base) {
        int res = type - base;
        return res >= 0 && res < 10;
    }

    public static boolean isKind(int type, int kind) {
        return type % BY_DATE == kind;
    }

    public static boolean isSame(int type, int base) {
        return type == base;
    }

    public static boolean isGpsType(int type) {
        return isBase(type, BY_LOCATION) || isBase(type, BY_OUT) || isBase(type, BY_PLACES);
    }

    public Reminder() {
        this.uuId = UUID.randomUUID().toString();
        this.uniqueId = new Random().nextInt(Integer.MAX_VALUE);
        this.isActive = true;
        this.isRemoved = false;
        this.useGlobal = true;
    }

    public Reminder(Reminder item, boolean fullCopy) {
        this.summary = item.getSummary();
        this.reminderType = item.getReminderType();
        this.groupUuId = item.getGroupUuId();
        this.eventCount = 0;
        this.color = item.getColor();
        this.delay = 0;
        this.vibrate = item.isVibrate();
        this.repeatNotification = item.isRepeatNotification();
        this.notifyByVoice = item.isNotifyByVoice();
        this.awake = item.isAwake();
        this.unlock = item.isUnlock();
        this.exportToTasks = item.isExportToTasks();
        this.exportToCalendar = item.isExportToCalendar();
        this.useGlobal = item.isUseGlobal();
        this.from = item.getFrom();
        this.to = item.getTo();
        this.hours = item.getHours();
        this.fileName = item.getFileName();
        this.melodyPath = item.getMelodyPath();
        this.volume = item.getVolume();
        this.dayOfMonth = item.getDayOfMonth();
        this.repeatInterval = item.getRepeatInterval();
        this.repeatLimit = item.getRepeatLimit();
        this.after = item.getAfter();
        this.weekdays = item.getWeekdays();
        this.type = item.getType();
        this.target = item.getTarget();
        this.subject = item.getSubject();
        this.attachmentFile = item.getAttachmentFile();
        this.attachmentFiles = item.getAttachmentFiles();
        this.auto = item.isAuto();
        this.isActive = item.isActive();
        this.isRemoved = item.isRemoved();
        this.isNotificationShown = item.isNotificationShown();
        this.isLocked = item.isLocked();
        this.places = item.getPlaces();
        this.shoppings = item.getShoppings();
        this.duration = item.getDuration();
        this.monthOfYear = item.getMonthOfYear();
        this.remindBefore = item.getRemindBefore();
        if (fullCopy) {
            this.uuId = item.getUuId();
            this.uniqueId = item.getUniqueId();
        }
    }

    public Reminder(RealmReminder item) {
        this.summary = item.getSummary();
        this.noteId = item.getNoteId();
        this.reminderType = item.getReminderType();
        this.groupUuId = item.getGroupUuId();
        this.uuId = item.getUuId();
        this.eventTime = item.getEventTime();
        this.startTime = item.getStartTime();
        this.eventCount = item.getEventCount();
        this.color = item.getColor();
        this.delay = item.getStatus();
        this.vibrate = item.isVibrate();
        this.repeatNotification = item.isRepeatNotification();
        this.notifyByVoice = item.isNotifyByVoice();
        this.awake = item.isAwake();
        this.unlock = item.isUnlock();
        this.exportToTasks = item.isExportToTasks();
        this.exportToCalendar = item.isExportToCalendar();
        this.useGlobal = item.isUseGlobal();
        this.from = item.getFrom();
        this.to = item.getTo();
        this.hours = wrapIntegerArray(item.getHours());
        this.fileName = item.getFileName();
        this.melodyPath = item.getMelodyPath();
        this.volume = item.getVolume();
        this.dayOfMonth = item.getDayOfMonth();
        this.repeatInterval = item.getRepeatInterval();
        this.repeatLimit = item.getRepeatLimit();
        this.after = item.getAfter();
        this.weekdays = wrapIntegerArray(item.getWeekdays());
        this.type = item.getType();
        this.target = item.getTarget();
        this.subject = item.getSubject();
        this.attachmentFile = item.getAttachmentFile();
        this.attachmentFiles = wrapStringArray(item.getAttachmentFiles());
        this.auto = item.isAuto();
        this.uniqueId = item.getUniqueId();
        this.isActive = item.isActive();
        this.isRemoved = item.isRemoved();
        this.isNotificationShown = item.isNotificationShown();
        this.isLocked = item.isLocked();
        this.duration = item.getDuration();
        this.monthOfYear = item.getMonthOfYear();
        this.remindBefore = item.getRemindBefore();
        this.places = new ArrayList<>();
        for (RealmPlace2 place : item.getPlaces()) {
            places.add(new Place(place));
        }
        this.shoppings = new ArrayList<>();
        for (RealmShopItem shopItem : item.getShoppings()) {
            shoppings.add(new ShopItem(shopItem));
        }
    }

    private List<Integer> wrapIntegerArray(List<RealmInteger> list) {
        List<Integer> strings = new ArrayList<>();
        if (list != null) {
            for (RealmInteger integer : list) {
                strings.add(integer.getInteger());
            }
        }
        return strings;
    }

    private List<String> wrapStringArray(List<RealmString> list) {
        List<String> strings = new ArrayList<>();
        if (list != null) {
            for (RealmString string : list) {
                strings.add(string.getString());
            }
        }
        return strings;
    }

    public Reminder copy() {
        Reminder reminder = new Reminder(this, false);
        reminder.setUuId(UUID.randomUUID().toString());
        reminder.setUniqueId(new Random().nextInt(Integer.MAX_VALUE));
        reminder.setActive(true);
        reminder.setRemoved(false);
        return reminder;
    }

    public long getRemindBefore() {
        return remindBefore;
    }

    public void setRemindBefore(long remindBefore) {
        this.remindBefore = remindBefore;
    }

    public int getMonthOfYear() {
        return monthOfYear;
    }

    public void setMonthOfYear(int monthOfYear) {
        this.monthOfYear = monthOfYear;
    }

    public long getDuration() {
        return duration;
    }

    public void setDuration(long duration) {
        this.duration = duration;
    }

    public boolean isNotificationShown() {
        return isNotificationShown;
    }

    public Reminder setNotificationShown(boolean notificationShown) {
        isNotificationShown = notificationShown;
        return this;
    }

    public boolean isLocked() {
        return isLocked;
    }

    public Reminder setLocked(boolean locked) {
        isLocked = locked;
        return this;
    }

    public boolean isActive() {
        return isActive;
    }

    public boolean isRemoved() {
        return isRemoved;
    }

    public Reminder setActive(boolean active) {
        isActive = active;
        this.isNotificationShown = false;
        this.isLocked = false;
        return this;
    }

    public Reminder setRemoved(boolean removed) {
        isRemoved = removed;
        return this;
    }

    public int getUniqueId() {
        return uniqueId;
    }

    public Reminder setUniqueId(int uniqueId) {
        this.uniqueId = uniqueId;
        return this;
    }

    public String getNoteId() {
        return noteId;
    }

    public Reminder setNoteId(String noteId) {
        this.noteId = noteId;
        return this;
    }

    public boolean isUseGlobal() {
        return useGlobal;
    }

    public Reminder setUseGlobal(boolean useGlobal) {
        this.useGlobal = useGlobal;
        return this;
    }

    public String getSummary() {
        return summary;
    }

    public Reminder setSummary(String summary) {
        this.summary = summary;
        return this;
    }

    public int getReminderType() {
        return reminderType;
    }

    public Reminder setReminderType(int reminderType) {
        this.reminderType = reminderType;
        return this;
    }

    public String getGroupUuId() {
        return groupUuId;
    }

    public Reminder setGroupUuId(String groupUuId) {
        this.groupUuId = groupUuId;
        return this;
    }

    public String getUuId() {
        return uuId;
    }

    public Reminder setUuId(String uuId) {
        this.uuId = uuId;
        return this;
    }

    public String getEventTime() {
        return eventTime;
    }

    public Reminder setEventTime(String eventTime) {
        this.eventTime = eventTime;
        return this;
    }

    public String getStartTime() {
        return startTime;
    }

    public Reminder setStartTime(String startTime) {
        this.startTime = startTime;
        return this;
    }

    public long getEventCount() {
        return eventCount;
    }

    public Reminder setEventCount(long eventCount) {
        this.eventCount = eventCount;
        return this;
    }

    public int getColor() {
        return color;
    }

    public Reminder setColor(int color) {
        this.color = color;
        return this;
    }

    public int getDelay() {
        return delay;
    }

    public Reminder setDelay(int delay) {
        this.delay = delay;
        return this;
    }

    public boolean isVibrate() {
        return vibrate;
    }

    public Reminder setVibrate(boolean vibrate) {
        this.vibrate = vibrate;
        return this;
    }

    public boolean isRepeatNotification() {
        return repeatNotification;
    }

    public Reminder setRepeatNotification(boolean repeatNotification) {
        this.repeatNotification = repeatNotification;
        return this;
    }

    public boolean isNotifyByVoice() {
        return notifyByVoice;
    }

    public Reminder setNotifyByVoice(boolean notifyByVoice) {
        this.notifyByVoice = notifyByVoice;
        return this;
    }

    public boolean isAwake() {
        return awake;
    }

    public Reminder setAwake(boolean awake) {
        this.awake = awake;
        return this;
    }

    public boolean isUnlock() {
        return unlock;
    }

    public Reminder setUnlock(boolean unlock) {
        this.unlock = unlock;
        return this;
    }

    public boolean isExportToTasks() {
        return exportToTasks;
    }

    public Reminder setExportToTasks(boolean exportToTasks) {
        this.exportToTasks = exportToTasks;
        return this;
    }

    public boolean isExportToCalendar() {
        return exportToCalendar;
    }

    public Reminder setExportToCalendar(boolean exportToCalendar) {
        this.exportToCalendar = exportToCalendar;
        return this;
    }

    public String getFrom() {
        return from;
    }

    public Reminder setFrom(String from) {
        this.from = from;
        return this;
    }

    public String getTo() {
        return to;
    }

    public Reminder setTo(String to) {
        this.to = to;
        return this;
    }

    public List<Integer> getHours() {
        return hours;
    }

    public Reminder setHours(List<Integer> hours) {
        this.hours = hours;
        return this;
    }

    public String getFileName() {
        return fileName;
    }

    public Reminder setFileName(String fileName) {
        this.fileName = fileName;
        return this;
    }

    public String getMelodyPath() {
        return melodyPath;
    }

    public Reminder setMelodyPath(String melodyPath) {
        this.melodyPath = melodyPath;
        return this;
    }

    public int getVolume() {
        return volume;
    }

    public Reminder setVolume(int volume) {
        this.volume = volume;
        return this;
    }

    public int getDayOfMonth() {
        return dayOfMonth;
    }

    public Reminder setDayOfMonth(int dayOfMonth) {
        this.dayOfMonth = dayOfMonth;
        return this;
    }

    public long getRepeatInterval() {
        return repeatInterval;
    }

    public Reminder setRepeatInterval(long repeatInterval) {
        this.repeatInterval = repeatInterval;
        return this;
    }

    public int getRepeatLimit() {
        return repeatLimit;
    }

    public Reminder setRepeatLimit(int repeatLimit) {
        this.repeatLimit = repeatLimit;
        return this;
    }

    public long getAfter() {
        return after;
    }

    public Reminder setAfter(long after) {
        this.after = after;
        return this;
    }

    public List<Integer> getWeekdays() {
        return weekdays;
    }

    public Reminder setWeekdays(List<Integer> weekdays) {
        this.weekdays = weekdays;
        return this;
    }

    public int getType() {
        return type;
    }

    public Reminder setType(int type) {
        this.type = type;
        return this;
    }

    public String getTarget() {
        return target;
    }

    public Reminder setTarget(String target) {
        this.target = target;
        return this;
    }

    public String getSubject() {
        return subject;
    }

    public Reminder setSubject(String subject) {
        this.subject = subject;
        return this;
    }

    public String getAttachmentFile() {
        return attachmentFile;
    }

    public Reminder setAttachmentFile(String attachmentFile) {
        this.attachmentFile = attachmentFile;
        return this;
    }

    public List<String> getAttachmentFiles() {
        return attachmentFiles;
    }

    public Reminder setAttachmentFiles(List<String> attachmentFiles) {
        this.attachmentFiles = attachmentFiles;
        return this;
    }

    public boolean isAuto() {
        return auto;
    }

    public Reminder setAuto(boolean auto) {
        this.auto = auto;
        return this;
    }

    public List<Place> getPlaces() {
        return places;
    }

    public Reminder setPlaces(List<Place> places) {
        this.places = places;
        return this;
    }

    public List<ShopItem> getShoppings() {
        return shoppings;
    }

    public Reminder setShoppings(List<ShopItem> shoppings) {
        this.shoppings = shoppings;
        return this;
    }

    public void setClear(ReminderInterface mInterface) {
        setSummary(mInterface.getSummary());
        setGroupUuId(mInterface.getGroup());
        setRepeatLimit(mInterface.getRepeatLimit());
        setColor(mInterface.getLedColor());
        setMelodyPath(mInterface.getMelodyPath());
        setVolume(mInterface.getVolume());
        setAuto(mInterface.getAuto());
        setAttachmentFile(mInterface.getAttachment());
        setActive(true);
        setRemoved(false);
        setDelay(0);
        setEventCount(0);
        setVibrate(mInterface.getVibration());
        setNotifyByVoice(mInterface.getVoice());
        setRepeatNotification(mInterface.getNotificationRepeat());
        setUseGlobal(mInterface.getUseGlobal());
        setUnlock(mInterface.getUnlock());
        setAwake(mInterface.getWake());
    }

    public long getDateTime() {
        return TimeUtil.getDateTimeFromGmt(eventTime);
    }

    public long getStartDateTime() {
        return TimeUtil.getDateTimeFromGmt(startTime);
    }

    @Override
    public int getViewType() {
        if (isSame(type, BY_DATE_SHOP)) {
            return SHOPPING;
        } else {
            return REMINDER;
        }
    }

    @Override
    public String toString() {
        return SuperUtil.getObjectPrint(this, Reminder.class);
    }

    @Override
    public int hashCode() {
        return UUID.fromString(uuId).hashCode();
    }

    public int[] getKeys() {
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(TimeUtil.getDateTimeFromGmt(eventTime));
        int y = calendar.get(Calendar.YEAR);
        int m = calendar.get(Calendar.MONTH);
        int d = calendar.get(Calendar.DAY_OF_MONTH);
        int h = calendar.get(Calendar.HOUR_OF_DAY);
        int min = calendar.get(Calendar.MINUTE);
        return new int[]{y, m, d, h, min};
    }

    public static class Kind {
        public static final int SMS = 2;
        public static final int CALL = 1;
    }
}

File: app/src/main/java/com/elementary/tasks/core/data/models/Reminder.kt
package com.elementary.tasks.core.data.models

import android.os.Parcelable
import androidx.annotation.Keep
import androidx.room.*
import com.elementary.tasks.core.data.converters.ListIntTypeConverter
import com.elementary.tasks.core.data.converters.ListStringTypeConverter
import com.elementary.tasks.core.data.converters.PlacesTypeConverter
import com.elementary.tasks.core.data.converters.ShopItemsTypeConverter
import com.elementary.tasks.core.interfaces.RecyclerInterface
import com.elementary.tasks.core.utils.datetime.DateTimeManager
import com.google.gson.annotations.SerializedName
import kotlinx.parcelize.Parcelize
import java.util.*

@Entity
@TypeConverters(
  PlacesTypeConverter::class,
  ShopItemsTypeConverter::class,
  ListStringTypeConverter::class,
  ListIntTypeConverter::class
)
@Keep
@Parcelize
data class Reminder(
  @SerializedName("summary")
  var summary: String = "",
  @SerializedName("noteId")
  var noteId: String = "",
  @SerializedName("reminderType")
  var reminderType: Int = 0,
  @SerializedName("eventState")
  var eventState: Int = 1,
  @SerializedName("groupUuId")
  var groupUuId: String = "",
  @SerializedName("uuId")
  @PrimaryKey
  var uuId: String = UUID.randomUUID().toString(),
  @SerializedName("eventTime")
  var eventTime: String = "",
  @SerializedName("startTime")
  var startTime: String = "",
  @SerializedName("eventCount")
  var eventCount: Long = 0,
  @SerializedName("color")
  var color: Int = 0,
  @SerializedName("delay")
  var delay: Int = 0,
  @SerializedName("vibrate")
  var vibrate: Boolean = false,
  @SerializedName("repeatNotification")
  var repeatNotification: Boolean = false,
  @SerializedName("notifyByVoice")
  var notifyByVoice: Boolean = false,
  @SerializedName("awake")
  @Deprecated("Not supported in newer OS versions")
  var awake: Boolean = false,
  @SerializedName("unlock")
  var unlock: Boolean = false,
  @SerializedName("exportToTasks")
  var exportToTasks: Boolean = false,
  @SerializedName("exportToCalendar")
  var exportToCalendar: Boolean = false,
  @SerializedName("useGlobal")
  var useGlobal: Boolean = true,
  @SerializedName("from")
  var from: String = "",
  @SerializedName("to")
  var to: String = "",
  @SerializedName("hours")
  var hours: List<Int> = ArrayList(),
  @SerializedName("fileName")
  var fileName: String = "",
  @SerializedName("melodyPath")
  var melodyPath: String = "",
  @SerializedName("volume")
  var volume: Int = -1,
  @SerializedName("dayOfMonth")
  var dayOfMonth: Int = 0,
  @SerializedName("monthOfYear")
  var monthOfYear: Int = 0,
  @SerializedName("repeatInterval")
  var repeatInterval: Long = 0,
  @SerializedName("repeatLimit")
  var repeatLimit: Int = -1,
  @SerializedName("after")
  var after: Long = 0,
  @SerializedName("weekdays")
  var weekdays: List<Int> = ArrayList(),
  @SerializedName("type")
  var type: Int = 0,
  @SerializedName("target")
  var target: String = "",
  @SerializedName("subject")
  var subject: String = "",
  @SerializedName("attachmentFile")
  @Deprecated("Use #attachmentFiles property")
  var attachmentFile: String = "",
  @SerializedName("attachmentFiles")
  var attachmentFiles: List<String> = ArrayList(),
  @SerializedName("auto")
  var auto: Boolean = false,
  @SerializedName("places")
  var places: List<Place> = ArrayList(),
  @SerializedName("shoppings")
  var shoppings: List<ShopItem> = ArrayList(),
  @SerializedName("uniqueId")
  var uniqueId: Int = Random().nextInt(Integer.MAX_VALUE),
  @SerializedName("isActive")
  var isActive: Boolean = true,
  @SerializedName("isRemoved")
  var isRemoved: Boolean = false,
  @SerializedName("isNotificationShown")
  var isNotificationShown: Boolean = false,
  @SerializedName("isLocked")
  var isLocked: Boolean = false,
  @SerializedName("hasReminder")
  var hasReminder: Boolean = false,
  @SerializedName("duration")
  var duration: Long = 0,
  @SerializedName("calendarId")
  var calendarId: Long = 0,
  @SerializedName("remindBefore")
  var remindBefore: Long = 0,
  @SerializedName("windowType")
  var windowType: Int = 0,
  @SerializedName("priority")
  var priority: Int = 2,
  @SerializedName("updatedAt")
  var updatedAt: String? = null,
  @SerializedName("taskListId")
  var taskListId: String? = null,
  @SerializedName("recurDataObject")
  var recurDataObject: String? = null,
  @ColumnInfo(name = "groupTitle")
  @Transient
  var groupTitle: String? = "",
  @ColumnInfo(name = "groupColor")
  @Transient
  var groupColor: Int = 0
) : RecyclerInterface, Parcelable {

  override val viewType: Int
    get() = if (isSame(type, BY_DATE_SHOP)) {
      SHOPPING
    } else {
      REMINDER
    }

  @Ignore
  constructor(item: Reminder, fullCopy: Boolean) : this() {
    this.summary = item.summary
    this.reminderType = item.reminderType
    this.groupUuId = item.groupUuId
    this.eventCount = 0
    this.color = item.color
    this.delay = 0
    this.vibrate = item.vibrate
    this.repeatNotification = item.repeatNotification
    this.notifyByVoice = item.notifyByVoice
    this.awake = item.awake
    this.unlock = item.unlock
    this.exportToTasks = item.exportToTasks
    this.exportToCalendar = item.exportToCalendar
    this.useGlobal = item.useGlobal
    this.from = item.from
    this.to = item.to
    this.hours = item.hours
    this.fileName = item.fileName
    this.melodyPath = item.melodyPath
    this.volume = item.volume
    this.dayOfMonth = item.dayOfMonth
    this.repeatInterval = item.repeatInterval
    this.repeatLimit = item.repeatLimit
    this.after = item.after
    this.weekdays = item.weekdays
    this.type = item.type
    this.target = item.target
    this.subject = item.subject
    this.attachmentFile = item.attachmentFile
    this.attachmentFiles = item.attachmentFiles
    this.auto = item.auto
    this.isActive = item.isActive
    this.isRemoved = item.isRemoved
    this.isNotificationShown = item.isNotificationShown
    this.isLocked = item.isLocked
    this.places = item.places
    this.shoppings = item.shoppings
    this.duration = item.duration
    this.monthOfYear = item.monthOfYear
    this.remindBefore = item.remindBefore
    this.windowType = item.windowType
    this.priority = item.priority
    this.hasReminder = item.hasReminder
    this.groupTitle = item.groupTitle
    this.groupColor = item.groupColor
    this.calendarId = item.calendarId
    this.taskListId = item.taskListId
    if (fullCopy) {
      this.uuId = item.uuId
      this.uniqueId = item.uniqueId
      this.updatedAt = item.updatedAt
    } else {
      this.uuId = UUID.randomUUID().toString()
      this.uniqueId = Random().nextInt(Integer.MAX_VALUE)
      this.updatedAt = DateTimeManager.gmtDateTime
    }
  }

  fun isLimited(): Boolean = repeatLimit > 0

  fun isLimitExceed(): Boolean = (repeatLimit - eventCount - 1) < 0

  fun copy(): Reminder {
    val reminder = Reminder(this, false)
    reminder.uuId = UUID.randomUUID().toString()
    reminder.uniqueId = Random().nextInt(Integer.MAX_VALUE)
    reminder.isActive = true
    reminder.isRemoved = false
    return reminder
  }

  fun isRepeating(): Boolean {
    return !isGpsType(type) && (repeatInterval > 0L || isBase(type, BY_WEEK)
      || isBase(type, BY_MONTH) || isBase(type, BY_DAY_OF_YEAR))
  }

  object Kind {
    const val SMS = 2
    const val CALL = 1
  }

  companion object {

    const val REMINDER = 0
    const val SHOPPING = 1

    const val BY_DATE = 10
    const val BY_DATE_CALL = 11
    const val BY_DATE_SMS = 12
    const val BY_DATE_APP = 13
    const val BY_DATE_LINK = 14
    const val BY_DATE_SHOP = 15
    const val BY_DATE_EMAIL = 16
    const val BY_TIME = 20
    const val BY_TIME_CALL = 21
    const val BY_TIME_SMS = 22
    const val BY_WEEK = 30
    const val BY_WEEK_CALL = 31
    const val BY_WEEK_SMS = 32
    const val BY_LOCATION = 40
    const val BY_LOCATION_CALL = 41
    const val BY_LOCATION_SMS = 42
    const val BY_MONTH = 60
    const val BY_MONTH_CALL = 61
    const val BY_MONTH_SMS = 62
    const val BY_OUT = 70
    const val BY_OUT_CALL = 71
    const val BY_OUT_SMS = 72
    const val BY_PLACES = 80
    const val BY_PLACES_CALL = 81
    const val BY_PLACES_SMS = 82
    const val BY_DAY_OF_YEAR = 90
    const val BY_DAY_OF_YEAR_CALL = 91
    const val BY_DAY_OF_YEAR_SMS = 92
    const val BY_RECUR = 100
    const val BY_RECUR_CALL = 101
    const val BY_RECUR_SMS = 102

    fun gpsTypes(): IntArray {
      return intArrayOf(
        BY_LOCATION,
        BY_LOCATION_CALL,
        BY_LOCATION_SMS,
        BY_OUT,
        BY_OUT_CALL,
        BY_OUT_SMS,
        BY_PLACES,
        BY_PLACES_CALL,
        BY_PLACES_SMS
      )
    }

    fun isBase(type: Int, base: Int): Boolean {
      val res = type - base
      return res in 0..9
    }

    fun isKind(type: Int, kind: Int): Boolean {
      return type % BY_DATE == kind
    }

    fun isSame(type: Int, base: Int): Boolean {
      return type == base
    }

    fun isGpsType(type: Int): Boolean {
      return isBase(type, BY_LOCATION) || isBase(type, BY_OUT) || isBase(type, BY_PLACES)
    }
  }
}
